### DISCUSSION TITLE
Testing the Cookie Consent Banner

### DISCUSSION THREAD

**Discussion by arakoczy-r2o** [orestbida/cookieconsent#512] [Q&A] [answered]
Hey there 👋 

First of all, thank you for this awesome project We've recently introduced it and it is really cool!

I'd have a question though, and I'm sorry if it was already asked or described before, I haven't found anything regarding it.

As we track the user's interaction on the banner we have a `tracking` function call on both `onFirstAction` and `onChange` events with different names. It works like a charm, but when we were trying to write consecutive tests for the `tracking` function call,  we hit the problem that the module doesn't seem to be "refreshing" in different renders.

```
it("called once and not on rerenders", async () => {
      const { rerender, findByText } = renderComponent();

      await acceptPreferences(findByText);
      expect(trackingMock).toHaveBeenCalledTimes(1);
      // initial event is 'updateSettings'
      expect(trackingMock).toHaveBeenCalledWith({
        event: "updateSettings",
      });
      rerenderWithIntlProvider(rerender, <CookieConsentBanner />);
      expect(trackingMock).toHaveBeenCalledTimes(1);
    });

it("is called with when user changes settings ", async () => {
      const { findByText } = renderComponent();

      openCookieSettings();

      // subsequent event is 'changeSettings'
      await acceptAll(findByText);
      expect(trackingMock).toHaveBeenCalledTimes(1);
      expect(trackingMock).toHaveBeenLastCalledWith({
        event: "changeSettings",
      });
    });
```

In the second test case, we cannot test the `trackingMock` to be called with `updateSettings` (onFirstAction), but only with `changeSettings` (onChange). Do you have any idea how to solve this?

Package version
```
vanilla-cookieconsent": "2.8.9"
```

Thank you in advance!

**Comment by orestbida** [comment, marked as answer]
Hi @arakoczy-r2o,

you must manually delete the root `'cc--main'` element and also the cookie `'cc_cookie'` for the plugin to reset.

`onFirstAction` is executed only if consent has yet to be expressed, or in other terms, if `cc_cookie` is invalid.

To address this usecase, a [`.reset()`](https://cookieconsent.orestbida.com/reference/api-reference.html#reset) method was introduced in v3.

**Reply by arakoczy-r2o** [reply]
Awesome, thank you very much!
